src/sctools/test/test_count.py

```diff
@@ -1,44 +1,578 @@
+                            + len(incomplete_alignment_tags_list)
+                            + len(multiple_alignment_tags_list))
+
+    def _get_query_name(self, query_prefix: str, i_query: int) -> str:
+        num_digits = len(str(self.num_queries - 1))
+        return query_prefix + str(i_query).zfill(num_digits)
+
+    def _generate_aligned_segment_from_tags(self,
+                                            alignment_tags: AlignmentRecordTags,
+                                            reference_start: int,
+                                            query_prefix: str,
+                                            i_query: int) -> pysam.AlignedSegment:
+        """Generates pysam.AlignedSegment instances from the contained tags.
+
+        Parameters
+        ----------
+        alignment_tags : AlignmentRecordTags
+            tags to attach to the instantiated pysam.AlignedSegment
+        reference_start : int
+            start position on the synthetic reference
+        query_prefix : str
+            prefix to use for query name
+        i_query : int
+            query index
+
+        Notes
+        -----
+        The query_sequence and query_quality are both empty as these features are not used for generating
+        the counts matrix. Likewise, the flag is currently unset. In the future, once we add a filtering
+        policy based on BAM record flags (such as duplicates), this method must be updated accordingly.
+
+        Returns
+        -------
+        pysam.AlignedSegment
+            an instance of pysam.AlignedSegment
+
+        """
+        tags = []
+        if alignment_tags.cell_barcode:
+            tags.append((SyntheticBarcodedBAMGenerator.cell_barcode_tag,
+                         alignment_tags.cell_barcode, 'Z'))
+        if alignment_tags.molecule_barcode:
+            tags.append((SyntheticBarcodedBAMGenerator.molecule_barcode_tag,
+                         alignment_tags.molecule_barcode, 'Z'))
+        if alignment_tags.gene_name:
+            tags.append((SyntheticBarcodedBAMGenerator.gene_name_tag,
+                         alignment_tags.gene_name, 'Z'))
+        record = pysam.AlignedSegment()
+        record.query_name = self._get_query_name(query_prefix, i_query)
+        record.reference_start = reference_start
+        record.reference_id = 0  # note: we only use one synthetic sequence
+        if len(tags) > 0:
+            record.set_tags(tags)
+        return record
+
+    def get_bam_records_generator(self, rng_seed: int = 777) -> Generator[pysam.AlignedSegment, None, None]:
+        """Returns a generator of pysam.AlignedSegment instances created from the alignment tags
+        provided to the initializer."""
+        rng = np.random.RandomState(rng_seed)
+        i_query = 0
+
+        # primary alignments
+        for alignment_tags in self.primary_alignment_record_tags_list:
+            reference_start = rng.randint(low=0, high=self.synthetic_sequence_length)
```

Comment:
At the moment, `sequence_name` and `sequence_start` are completely random and arbitrary (since we are not using these alignment features for counting -- only tags). Perhaps it makes sense if we go ahead and generate appropriate BAM header from hg38 reference and `sequence_name` and `sequence_start` and CIGAR strings that actually to an alignment to the gene shown in the tag. I would, however, willing to do this if we plan to feed the BAM file to cellranger (or use these features in counting). Let's discuss.

Reply:
I will add additional docblocks regarding the nature of the test data.

Reply:
Ah got it, you're query name sorting, so you don't need the alignment information? 

One part of me wishes we'd just fill with 0/nan/none/NNNN or something very obviously unused, but I can see the advantage of putting the right machinery in place to generate real sequences for more robust testing later. 

More docblocks seems like the right approach here. 👍